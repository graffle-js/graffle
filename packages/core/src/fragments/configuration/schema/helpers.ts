/**
 * Extract schema name from context for multi-schema LSP support.
 *
 * Falls back to `never` if no schema name is configured.
 */
// dprint-ignore
export type Name<$Context> =
  $Context extends { configuration: { schema: { current: { name: infer $Name } } } }
    ? $Name
    : never

/**
 * Check if context has SDDM (Schema-Driven Data Map) configured.
 *
 * Returns `true` if schema.map is defined and not undefined, `false` otherwise.
 */
// dprint-ignore
export type HasMap<$Context> =
  $Context extends { configuration: { schema: { current: { map: infer $Map } } } }
    ? $Map extends undefined ? false : true
    : false

/**
 * Extract schema information from context.
 *
 * Returns an object containing the schema and map from the GlobalRegistry,
 * or `never` if the schema is not properly configured.
 */
// dprint-ignore
export type Info<$Context> =
  $Context extends {
    configuration: {
      schema: {
        current: {
          name: infer $Name extends string
          map: infer $Map
        }
      }
    }
  }
    ? $Name extends keyof GraffleGlobal.Clients
      ? {
          schema: GraffleGlobal.Clients[$Name]['schema']
          map: $Map
        }
      : never
    : never
