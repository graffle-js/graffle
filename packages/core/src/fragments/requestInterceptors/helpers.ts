import type { Configurator } from '@wollybeard/kit'
import type { Client_justContext, ClientEmpty } from '#~/client/client.js'
import type { RequestPipeline } from '#~/requestPipeline/RequestPipeline.js'
import { Ware } from '@wollybeard/kit'
import type { Context } from '../../context.js'
import type { ContextComputerParameters } from '../types.js'

// Type complexity simplified - actual types enforced in client package
export const create = <$Client extends Client_justContext = ClientEmpty>(
  interceptor: any,
) => interceptor

export type RequestInterceptorComputer<
  $Context extends Context = Context,
  $Configuration extends Configurator.Configuration = Configurator.Configuration,
> = (
  parameters: ContextComputerParameters<$Context, $Configuration>,
) => RequestPipeline.BaseInterceptor
