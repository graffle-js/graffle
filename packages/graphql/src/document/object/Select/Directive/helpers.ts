import { Include } from './include.js'
import { Skip } from './skip.js'

export const definitionsByName = {
  include: Include,
  skip: Skip,
}

export const keyPrefix = `$`

export const parseKey = (key: string): string | null => {
  if (key.startsWith(keyPrefix)) {
    return key.slice(keyPrefix.length)
  }

  return null
}
