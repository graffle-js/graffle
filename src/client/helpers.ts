import type { PartialOrUndefined } from '../lib/prelude.js'
import type { Context } from '../types/context.js'
import type { ClientGeneric } from './client.js'

export const createProperties = (
  callback: (
    clientConstructor: (context: Context) => ClientGeneric,
    context: Context,
  ) => PartialOrUndefined<ClientGeneric>,
) => {
  return callback
}
