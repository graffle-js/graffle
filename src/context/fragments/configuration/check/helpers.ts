import type { Bool } from '@wollybeard/kit'
import type { Configurator } from './check.js'

/**
 * Check if preflight checks are disabled in the context.
 *
 * Returns `true` if preflight is explicitly set to false, `false` otherwise.
 * This is the inverse of {@link IsPreflightEnabled}.
 */
// dprint-ignore
export type IsPreflightDisabled<$Context> = Bool.not<IsPreflightEnabled<$Context>>

/**
 * Check if preflight checks are enabled in the context.
 *
 * Returns `false` if preflight is explicitly set to false, `true` otherwise.
 * This is the inverse of {@link IsPreflightDisabled}.
 */
// dprint-ignore
export type IsPreflightEnabled<$Context> =
  $Context extends { configuration: { check: { current: { preflight: infer $Preflight extends boolean } } } }
    ? $Preflight
    : Configurator['default']['preflight']
