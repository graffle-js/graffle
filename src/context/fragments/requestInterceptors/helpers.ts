import type { Client_justContext, ClientEmpty } from '../../../client/client.js'
import type { Configurator } from '../../../entrypoints/extension_exports.js'
import type { Anyware } from '../../../lib/anyware/_namespace.js'
import type { RequestPipeline } from '../../../requestPipeline/RequestPipeline.js'
import type { Context } from '../../context.js'
import type { ContextComputerParameters } from '../types.js'

export const create = <$Client extends Client_justContext = ClientEmpty>(
  interceptor: Anyware.Interceptor.InferFromPipeline<
    Anyware.Pipeline.InferFromDefinition<$Client['_']['requestPipelineDefinition']>
  >,
) => interceptor

export type RequestInterceptorComputer<
  $Context extends Context = Context,
  $Configuration extends Configurator.Configuration = Configurator.Configuration,
> = (
  parameters: ContextComputerParameters<$Context, $Configuration>,
) => RequestPipeline.BaseInterceptor
