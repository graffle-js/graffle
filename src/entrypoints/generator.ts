export { generateFiles as generate } from '../layers/4_generator/files.js'
