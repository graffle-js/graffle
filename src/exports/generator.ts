export * from './__Generator.js'
