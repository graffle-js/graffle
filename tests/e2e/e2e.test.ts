import { expect } from 'vitest'
import { test } from '../_/helpers.js'

// - new project no generation
// 	- type error if try give name to constructor

// - new project with 1 client generation
// 	- type error if try give name to constructor
// 	- can use static client
// 	- can use generated client import

// - new project with 2 client generation
// 	- type error if give wrong name to constructor
// 	- can give right name to constructor
// 	- can use static client
// 	- can use generated clients import

// ...?

test(`client works without generation`, async ({ project, pokemonService }) => {
  await project.fs.writeAsync(
    `main.ts`,
    `
		import { Graffle } from 'graffle'
		const graffle = Graffle.create().transport({ url: '${pokemonService.url.href}' })
		const data = await graffle.gql\`
      query ($name: String!) {
        pokemonByName (name: $name) {
          name
          hp
          attack
          defense
          trainer {
            name
          }
        }
      }
    \`.send({ name: 'Pikachu' })
		console.log(data?.['pokemonByName'])
		`,
  )
  const result = await project.run`pnpm tsx main`
  expect(result.stdio).toMatchSnapshot()
})

test(`client works with generation`, async ({ project, pokemonService }) => {
  await project.fs.writeAsync(
    `main.ts`,
    `
		import { Graffle } from 'graffle'
		const graffle = Graffle.create().transport({ url: '${pokemonService.url.href}' })
		const data = await graffle.query.pokemonByName({
			$: { name: 'Pikachu' },
			name: true,
			hp: true,
			attack: true,
			defense: true,
			trainer: {
				name: true,
			},
		})
		console.log(data)
		`,
  )
  {
    const result = await project.run`pnpm graffle --schema http://localhost:3001/graphql`
    expect(result.stdio).toMatchSnapshot()
  }
  {
    await project.run`pnpm check:types`
  }
  {
    const result = await project.run`pnpm tsx main`
    expect(result.stdio).toMatchSnapshot()
  }
})

test(`client uses dprint formatter if installed`, async ({ project }) => {
  await project.addDprintConfig()
  const path = await project.addPokemonSchemaSDL()

  await project.run`pnpm add --save-dev dprint @dprint/formatter @dprint/typescript`

  const genResult = await project.run`pnpm graffle --schema ${path.relative} --format`
  const genResultStdout = genResult.stdout as string
  expect(genResultStdout.includes(`No TypeScript formatter found`)).toEqual(false)

  await project.run`pnpm dprint check graffle/**/*`
})

test(`client uses prettier formatter if installed`, async ({ project }) => {
  // await project.addDprintConfig()
  const path = await project.addPokemonSchemaSDL()

  await project.run`pnpm add --save-dev prettier`

  const genResult = await project.run`pnpm graffle --schema ${path.relative} --format`
  const genResultStdout = genResult.stdout as string
  expect(genResultStdout.includes(`No TypeScript formatter found`)).toEqual(false)

  await project.run`pnpm prettier --check graffle/**/*`
})

test(`project uses graffle config file if present`, async ({ project, onTestFinished }) => {
  // Clean up global side effects caused by this e2e test in development.
  const isCI = process.env[`CI`]
  if (!isCI) {
    onTestFinished(async () => {
      // If tests worked then this is already removed so catch expected error.
      await project.run(`pnpm`, [`-g`, `remove`, `tsx`]).catch(() => {})
      await project.run(`pnpm`, [`env`, `use`, `--global`, `lts`])
    })
  }
  const path = await project.addPokemonSchemaSDL()
  const configFilePaths = {
    ts: `graffle.config.ts`,
    js: `graffle.config.js`,
  }
  const usingConfigFilePatterns = {
    ts: new RegExp(`using file config found at.*${configFilePaths.ts}`, `i`),
    js: new RegExp(`using file config found at.*${configFilePaths.js}`, `i`),
  }
  const configContent = `
    import { Generator } from 'graffle/generator'
    export default Generator.configure({
      schema: {
        type: 'sdlFile',
        dirOrFilePath: '${path.relative}',
      },
    })
  `
  project.fs.write(configFilePaths.ts, configContent)

  {
    // TypeScript with local tsx (installed with project by default)
    const result = await project.run(`pnpm`, [`graffle`])
    expect(result.stdout).toMatch(usingConfigFilePatterns.ts)
  }
  {
    // JavaScript with local tsx
    project.fs.rename(configFilePaths.ts, configFilePaths.js)
    const result = await project.run(`pnpm`, [`graffle`])
    expect(result.stdout).toMatch(usingConfigFilePatterns.js)
  }
  {
    // TypeScript with global tsx
    await project.run(`pnpm`, [`remove`, `tsx`])
    project.fs.rename(configFilePaths.js, configFilePaths.ts)
    await project.run(`pnpm`, [`-g`, `add`, `tsx`])
    const result = await project.run(`pnpm`, [`graffle`])
    expect(result.stdout).toMatch(usingConfigFilePatterns.ts)
  }

  // Cover --experimental-strip-types
  await project.run(`pnpm`, [`-g`, `remove`, `tsx`])

  {
    // TypeScript with --experimental-strip-types
    const result = await project.run(`pnpm`, [`graffle`])
    expect(result.stdout).toMatch(usingConfigFilePatterns.ts)
  }
  {
    // JavaScript with --experimental-strip-types
    project.fs.rename(configFilePaths.ts, configFilePaths.js)
    const result = await project.run(`pnpm`, [`graffle`])
    expect(result.stdout).toMatch(usingConfigFilePatterns.js)
    expect(result.stderr).toMatch(/type Stripping is an experimental feature/i)
  }
  {
    // JavaScript with older Node.js version
    await project.run(`pnpm`, [`env`, `use`, `--global`, `22.5`])
    const result = await project.run(`pnpm`, [`graffle`])
    expect(result.stdout).toMatch(usingConfigFilePatterns.js)
    expect(result.stderr).not.toMatch(/type Stripping is an experimental feature/i)
  }
})
