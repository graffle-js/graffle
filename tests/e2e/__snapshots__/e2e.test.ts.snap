// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`client works with generation 1`] = `
[
  undefined,
  "WARNING: No TypeScript formatter found. Generated code will remain ugly. To have code automatically formatted do one of the following things:

- pnpm add --save-dev @dprint/formatter @dprint/typescript
- pnpm add --save-dev prettier

To suppress this warning disable formatting in one of the following ways:

- CLI: graffle --no-format
- Configuration file: Generator.configuration({ format: false })
- API: Generator.generate({ format: false })
WARNING: Custom scalars detected in the schema, but you have not created a custom scalars module to import implementations from.",
  "",
]
`;

exports[`client works with generation 2`] = `
[
  undefined,
  "[
  {
    name: 'Pikachu',
    hp: 35,
    attack: 55,
    defense: 40,
    trainer: { name: 'Ash' }
  }
]",
  "",
]
`;

exports[`client works without generation 1`] = `
[
  undefined,
  "[
  {
    name: 'Pikachu',
    hp: 35,
    attack: 55,
    defense: 40,
    trainer: { name: 'Ash' }
  }
]",
  "",
]
`;
